docs/serialization-design.md

```diff
@@ -0,0 +1,64 @@
+# Serialization
+
+DyND stores typed multi-dimensional data in memory, which needs to be saved to disk and transported between machines for things like distributed and out of core computation. We would like the serialization to be simple and obvious where possible, and well-defined in a way suitable for use as a building block for creating Merkle DAGs.
+
+What follows is a proposal for how this can be done, to get things started and provide a basis for design discussion.
+
+## Document/Design Scope
+
+Ultimately, we may want to add a streaming capability to DyND, so that a DyND array which is almost as big as memory could stream directly to a file without being placed in an intermediate bytes object. In initial design and implementation of serialization, we can simply use the `bytes` type without hampering such future extension.
+
+We are not concerned with deduplication, as that would complicate the serialization format a fair bit. Deduplication can be handled by a future extension to a new type like `array[Any]`, but which operates as a Merkle link, together with serialization to/from a content addressable storage (CAS).
+
+We are not concerned with compression, that is something that can be layered on top of serialization. If a stream interface is added to DyND, chaining compression to serialization via that interface would be natural to pursue.
+
+As a serialization protocol, this document is concerned with converting to/from byte streams, and  not with random access or efficient seeking. In some cases, random access is possible, however.
+
+## Serialize Callable
+
+The `serialize` callable has signature `(Any) -> bytes`, and operates as a reduction kernel. This means it concatenates data to the output bytes object. Only the data gets serialized, not the type. This is not a problem, however, because to serialize an arbitrary array, one can serialize it as `array[Any]`, which must serialize both the type and the data.
```

Comment:
Another reason, which I didn't write down, is that it has to be this way to be able to define the serialization properly. The type generally specifies the structure of the data quite rigidly, unlike JSON-like serialization formats where each element can have different types. If we put the type into the serialization of everything, we can't recursively apply that definition for serializing arrays, tuples, etc, without the type appearing repeatedly everywhere.